security/tutorials/calico-policy-advanced.md

```diff
@@ -0,0 +1,345 @@
+---
+title: Calico policy, advanced tutorial
+description: Learn how to create more advanced Calico network policies (namespace, allow and deny all ingress and egress).
+canonical_url: "/security/tutorials/calico-policy-advanced"
+---
+
+Calico Network Policies _extend_ the functionalities of Kubernetes Network Policies. To demonstrate this, this page will walk through the [Kubernetes Advanced NetworkPolicy](kubernetes-policy-advanced) tutorial but with everything implemented using Calico policies. This serves as a good example in highlighting the syntactical differences between the two type of policies while demonstrating Calico Network Policy's flexibility.
+
+### Requirements
+
+- Calico v2.6.1+ with Kubernetes 1.8+
+- A working Kubernetes cluster and access to it using kubectl and calicoctl
+- Your Kubernetes nodes have connectivity to the public internet
+- You are familiar with [Kubernetes NetworkPolicy](kubernetes-policy-basic) or [Kubernetes NetworkPolicy Advanced](kubernetes-policy-advanced)
+
+### Tutorial flow
+
+1. Create the Namespace and Nginx Service
+1. Deny all ingress traffic
+1. Allow ingress traffic to Nginx
+1. Deny all egress traffic
+1. Allow egress traffic to kube-dns
+1. Cleanup Namespace
+
+### 1. Create the namespace and nginx service
+
+We'll use a new namespace for this guide. Run the following commands to create it and a plain nginx service listening on port 80.
+
+```bash
+kubectl create ns advanced-policy-demo
+kubectl run --namespace=advanced-policy-demo nginx --replicas=2 --image=nginx
+kubectl expose --namespace=advanced-policy-demo deployment nginx --port=80
+```
+
+#### Verify access - allowed all ingress and egress
+
+Open up a second shell session which has `kubectl` connectivity to the Kubernetes cluster and create a busybox pod to test policy access. This pod will be used throughout this tutorial to test policy access.
+
+```bash
+kubectl run --namespace=advanced-policy-demo access --rm -ti --image busybox /bin/sh
+```
+
+This should open up a shell session inside the `access` pod, as shown below.
+
+```
+Waiting for pod advanced-policy-demo/access-472357175-y0m47 to be running, status is Pending, pod ready: false
+
+If you don't see a command prompt, try pressing enter.
+/ #
+```
+
+{: .no-select-button}
+
+Now from within the busybox "access" pod execute the following command to test access to the nginx service.
+
+```bash
+wget -q --timeout=5 nginx -O -
+```
+
+It should return the HTML of the nginx welcome page.
+
+Still within the busybox "access" pod, issue the following command to test access to google.com.
+
+```bash
+wget -q --timeout=5 google.com -O -
+```
+
+It should return the HTML of the google.com home page.
+
+### 2. Deny all ingress traffic
+
+Enable ingress isolation on the namespace by deploying a [default deny all ingress traffic policy](https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-deny-all-ingress-traffic).
```

Comment:
I suggest changing the link to be to the Calico docs page on default deny, and move the deny all steps to the end of the tutorial using GlobalNetworkPolicy so they apply across all namespaces (since this is something you can't do with k8s policy).

Reply:
Hey @lxpollitt I was wondering for your comment on moving the deny policy to the end of my tutorial I think it would be breaking the tutorial flow a little bit, I think the default deny all ingress traffic is there in the beginning of the the tutorial to showcase the next policies effects and so on, so it HAS to be at the beginning. I was thinking instead, of at the very end showcasing Calico's global deny (ingress / egress) policy to show that you can instead start off securing your workload by using Calico's GLOBAL policy instead of doing Kubernetes network policys and have everything secured so that only the traffic you define is allowed. What do you think? (I also moved the link to the calico documentation)

Reply:
One goal I was thinking of is for the policies the reader ends up with to reflect best practices in some way.  I feel like having separate policies for deny ingress, deny egress, and allow DNS, is probably not where most users would want to end up.  Maybe you could achieve this, and still have the build out steps you have today, by calling the policy "default-policy" and modifying it as you go along.  So the final policy is the "result" of all the learning.  WDYT?

Reply:
@lxpollitt I think that's a good way to go. I will build out that version of this tutorial and notify you and see what you think! :D